### PR TITLE
Enable notification badges

### PR DESCRIPTION
**Pull request by ibauersachs**
I'm noticing the missing notification counter badges mentioned in #1695 as well. This PR is based on the patch from @vaskion, but omits the animation changes.

### CODE REVIEW

**Comment by vaskion** [comment]
> 

Please keep in mind that in the latest **Ubuntu 22.10** for some applications the notifications (and the badges of course) will not go away automatically when you read them/open the app that posted them. I played around with my mods but was not able to fix it. I disabled the DashToPanel extension and it was still happening. So, I think this might be some issue with the OS itself as for these apps the notifications still stay in the system tray even you have read them until you click on the clock and then select 'Clear'. This is happening with Teams for me. Google Chrome and its chat app do not have this issue.

**Comment by Alt37** [comment]
@vaskion , @ibauersachs 
Thanks to the [RocketBar extension](https://github.com/linux-is-awesome/gnome_extension_rocketbar/blob/master/extension/utils/notificationService.js) I've managed to filter-out notification badges that won't go away until you clear them from the Message Tray (e.g. file operations on Nautilus), and some rather distracting ones  (e.g. "Window is ready").
Other applications I've tested which currently lack Unity API support (Hamsket/Rambox, Google Chrome, Firefox, Thunderbird) still display badges.

```
--- a/progress.js
+++ b/progress.js
@@ -392,7 +392,7 @@
             for (let i = 0; i < items.length; i++) {
                 let item = items[i];
                 // FIXME - Check if these 12 symbols are enough
-                if (this._source.app.get_name().includes(item.title.substring(0, 12))) {
+                if ( (this._source.app.get_name().includes(item.title.substring(0, 12))) && (item instanceof imports.ui.notificationDaemon.FdoNotificationDaemonSource) ) {
                     unreadCount += item.notifications.length;
                 }
             }
```

**Comment by Alt37** [comment]
I need to register Teams account for a test. On Ubuntu 22.04 for me "Window is ready" wasn't counted in any app yet (not sure how to trigger it with VSCodium though). Today few times had a problem with "Google Chat" PWA. Notification wasn't present on a message tray but badge still didn't go away until I cleared all notifications. Also once Google Chat PWA didn't receive badge until I deleted one of the notifications. Maybe some events got skipped, still testing... 
Can you please share your "progress.js" please? And could you try RocketBar extension? For "RocketBar" you'll need to select "keep original gnome-shell top panel" from "Dash-to-Panel" options or disable "Dash-to-Panel" extension)

**Comment by ibauersachs** [comment]
Note that the Teams app is deprecated, the "replacement" is just Edge with Teams as a PWA. Thus any website or PWA from Chrome/Chromium/Edge should expose the same behavior.

**Comment by Alt37** [comment]
For some reason
`this._source.app.id == item.app.get_id()`
stopped working for me so I've edited my post with a patch leaving only
`item instanceof imports.ui.notificationDaemon.FdoNotificationDaemonSource` as additional condition.

**Comment by vaskion** [comment]
> 
Yeah, try the 22.10. This happened after I installed the update. Maybe it is because I updated and did not do a clean install. Before that I have no issue with my original mod - it was working perfectly! I had to install the update though because I had some issues with the network which are now solved. So, it seems they fix something but break something else.

P.S. I am experience some other issues with this update as well - some buttons and menus are not rendered correctly and when I go with the mouse pointer over some items they got repainted and overlap with the others. Extremely ugly... This is the most immature and buggy OS I have ever seen so far...

**Comment by Yushi5058** [comment]
I'm waiting for this feature ..

**Comment by AndyNetDuma** [comment]
Can confirm this works well with 22.04. Can't comment on 22.10
Would be nice to get this merged so I don't have to manually install it from the branch

**Comment by kaimast** [comment]
Could you rebase this branch? Looks like it is over 100 commits behind upstream.

**Comment by xalt7x** [comment]
> But I need to address @xalt7x comment.

TLDR: Third-party developers might figure out some workaround but that's probably more of an upstream Chromium issue.

---

1. We've discussed this in [RocketBar issue #26](https://github.com/linux-is-awesome/gnome_extension_rocketbar/issues/26#issuecomment-1340814683) and extension author didn't figure out a simple solution.
2. With Dash to Dock v79 (which [introduced](https://github.com/micheleg/dash-to-dock/pull/1961) such notification badges) and newer this issue is reproducible as well.
3. Google Chrome Flatpak had [somewhat similar issue])(https://github.com/flathub/com.google.Chrome/issues/39) and now it uses Flextop wrapper to separate PWA launchers from the main Google Chrome (unfortunately this also breaks notification badges).

**Comment by ibauersachs** [comment]
@kaimast and @xalt7x 

I've updated the implementation to
- only trigger once for each notification
- properly compare the app id
- and to follow the Gnome policy, which is per app, if notifications should be shown or not

The issue that with Chrome/Edge is that the _first_ launched instance is always then one that sends the notifications. The app-id of the normal browser is e.g. `microsoft-edge.desktop` and for a PWA it's `msedge-faolnafnngnfdaknnbpnkhgohbobgegn-Default.desktop`, but that's not reflected in the notification. Whatever handles the notifications thus cannot distinguish between a regular browser window or a PWA.

**Comment by vaskion** [comment]
@ibauersachs Thanks a lot for looking into this!
I tested your changes and they work fine except for **MS Teams**. For some reason on Ubuntu **22.04** it does not clear the badge and the notification as well until you click on the clock and then the "Clear" button to remove all notifications. So, I am reverting back to stock variant. Unfortunately, these changes are not usable for me in their current form until MS Teams issue is fixed...

**Comment by ibauersachs** [comment]
@vaskion Is that the (deprecated) Teams app or via PWA?

**Comment by vaskion** [comment]
@ibauersachs I am not sure what you mean by PWA, but I am using a separate application I installed more than an year ago and not the web application inside Chrome or other browser. So, I guess it would be the deprecated thingy.

**Comment by xalt7x** [comment]
@vaskion 
Maybe you have issue similar to this?
https://github.com/ferdium/ferdium-app/issues/1142

**Comment by vaskion** [comment]
@xalt7x Yeah, very much seems like this one but it happens only with MS Teams so far. Anyways, I have given up from making the badge thing work on Ubuntu...

I thing I read somewhere that on Ubuntu 24.something the badges are already integrated in the OS itself but they ruined something else I use. Seems like one step forward, two steps back...

**Comment by brunologitravel** [comment]
is working in this one: https://github.com/micheleg/dash-to-dock

**Comment by NickBelhomme** [comment]
I tried this on the latest installed version on GNOME 46.
Does not work for me.
I do not see the notification indicator on the discord app.
It does show in the appIndicator extension...

image
